### DISCUSSION TITLE
Simple performance mesurement

### DISCUSSION THREAD

**Discussion by s5uishida** [edgecomllc/eupf#507] [General]
Hi,

It may not be very meaningful on Virtulabox VMs, but in order to evaluate the relative performance of several UPFs, I briefly measured the performance of UPFs.

https://github.com/s5uishida/simple_measurement_of_upf_performance

`xdp_attach_mode` was set to `generic`, but I expect that the performance will be significantly improved when set to `native` or `offload` for the real devices.

FYI.

**Comment by pirog-spb** [comment]
Hi @s5uishida 

Great job!

BTW `vertio-net` was mentioned. Kind of typo? Should be `virtio-net`? 

Why don't you configure `native` attach mode on eUPF during tests? Seems like virtio-net should support XDP. With the `native` mode these tests would be more indicative :)

**Reply by s5uishida** [reply]
Hi @pirog-spb 

Thank you very much for pointing this out. I corrected to `virtio-net`.

BTW, when I set the virtualbox VM's `virtio-net` to `native` and started it, the following error occurred.

```
# ./bin/eupf --config config.yml
2023/12/16 19:36:35 Get raw config: map[api_address::8080 echo_interval:10 far_map_size:1024 feature_ftup:true gtp_peer:[] heartbeat_interval:5 heartbeat_retries:3 heartbeat_timeout:5 interface_name:[enp0s9 enp0s16] logging_level:info metrics_address::9090 n3_address:192.168.13.151 pdr_map_size:1024 pfcp_address:192.168.14.151:8805 pfcp_node_id:192.168.14.151 qer_map_size:1024 resize_ebpf_maps:false teid_pool:65536 xdp_attach_mode:native]
2023/12/16 19:36:35 Apply eUPF config: {InterfaceName:[enp0s9 enp0s16] XDPAttachMode:native ApiAddress::8080 PfcpAddress:192.168.14.151:8805 PfcpNodeId:192.168.14.151 MetricsAddress::9090 N3Address:192.168.13.151 GtpPeer:[] EchoInterval:10 QerMapSize:1024 FarMapSize:1024 PdrMapSize:1024 EbpfMapResize:false HeartbeatRetries:3 HeartbeatInterval:5 HeartbeatTimeout:5 LoggingLevel:info FTEIDPool:65536 FeatureFTUP:true}
2023/12/16 19:36:35 FTL Could not attach XDP program: create link: operation not supported
```

If you know how to deal with this, I would be very grateful if you could let me know.

A related URL would also be very helpful.

Best regards,

--Shigeru

**Reply by pirog-spb** [reply]
Try to check this one https://github.com/torvalds/linux/commit/f600b690501550b94e83e07295d9c8b9c4c39f4e

**Reply by pirog-spb** [reply]
And this https://stackoverflow.com/questions/73990311/xdp-driver-mode-on-vm-interface

**Reply by pirog-spb** [reply]
And may be this https://en.opensuse.org/Testing_XDP_with_virtio-net

**Reply by s5uishida** [reply]
Hi @pirog-spb 

Thank you very much for your information.

I will try it.

**Reply by s5uishida** [reply]
Hi @pirog-spb 

I measured the performance with TRex and got very good results.

https://github.com/s5uishida/simple_measurement_of_upf_performance_6

FYI.

**Reply by pirog-spb** [reply]
Hi @s5uishida 

Thanks for info. Looks promising
